reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,84 +81,117 @@ object StreamSubscriber {
     case class UpstreamError(err: Throwable) extends State
 
     def step(in: Input): State => (State, F[Unit]) =
       in match {
         case OnSubscribe(s) => {
-          case RequestBeforeSubscription(req) => WaitingOnUpstream(s, req) -> F.delay(s.request(1))
-          case Uninitialized                  => Idle(s) -> F.unit
+          case RequestBeforeSubscription =>
+            Receiving(s) -> F.delay(s.request(maxDemand))
+          case Uninitialized =>
+            Idle(s) -> F.unit
+          case Idle(prev) => // we got a second subscription. cancel it
+            Idle(prev) -> F.delay(s.cancel)
+          case Receiving(prev) => // we got a second subscription. cancel it
+            Receiving(prev) -> F.delay(s.cancel)
+          case DownstreamCancellation =>
+            DownstreamCancellation -> F.unit
           case o =>
             val err = new Error(s"received subscription in invalid state [$o]")
             o -> (F.delay(s.cancel) >> F.raiseError(err))
         }
         case OnNext(a) => {
-          case WaitingOnUpstream(s, r) => Idle(s) -> r.complete(a.some.asRight)
-          case DownstreamCancellation  => DownstreamCancellation -> F.unit
-          case o                       => o -> F.raiseError(new Error(s"received record [$a] in invalid state [$o]"))
+          case Receiving(s) =>
+            Receiving(s) -> q.enqueue1(a.some.asRight)
+          case Idle(s) =>
+            Receiving(s) -> q.enqueue1(a.some.asRight)
+          case DownstreamCancellation =>
+            DownstreamCancellation -> F.unit
+          case o =>
+            o -> F.raiseError(new Error(s"received record [$a] in invalid state [$o]"))
         }
         case OnComplete => {
-          case WaitingOnUpstream(_, r) => UpstreamCompletion -> r.complete(None.asRight)
-          case _                       => UpstreamCompletion -> F.unit
+          case Receiving(_) | Idle(_) =>
+            UpstreamCompletion -> q.enqueue1(None.asRight)
+          case _ =>
+            UpstreamCompletion -> F.unit
         }
         case OnError(e) => {
-          case WaitingOnUpstream(_, r) => UpstreamError(e) -> r.complete(e.asLeft)
-          case _                       => UpstreamError(e) -> F.unit
+          case Receiving(_) | Idle(_) =>
+            UpstreamError(e) -> (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight))
+          case _ =>
+           UpstreamError(e) -> F.unit
         }
         case OnFinalize => {
-          case WaitingOnUpstream(sub, r) =>
-            DownstreamCancellation -> (F.delay(sub.cancel) >> r.complete(None.asRight))
-          case Idle(sub) => DownstreamCancellation -> F.delay(sub.cancel)
-          case o         => o -> F.unit
+          case Idle(sub) =>
+            DownstreamCancellation -> (F.delay(sub.cancel) >> q.enqueue1(None.asRight))
+          case Receiving(sub) =>
+            DownstreamCancellation -> (F.delay(sub.cancel) >> q.enqueue1(None.asRight))
+          case o =>
```

Comment:
Here we need to handle `Uninitialized` and `RequestBeforeSubscription` and transition into the `DownstreamCancellation` state.